docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+
+The proposed API for this functionality is described in code here:
+
+[//]: # (Got to remove the /venice prefix before pushing to main on linkedin/venice, otherwise links will be broken...)
+
+* [StoreLifecycleHooks](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleHooks.html), which is the main part of this 
+  proposal.
+* [StoreLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleEventOutcome.html), which is the signal
+  returned by some hooks to indicate that a given step should proceed or abort.
+* [StoreVersionLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreVersionLifecycleEventOutcome.html), which
+  is the signal returned by some other hooks which need more fine-grained control over the workflow. In addition to 
+  proceeding and aborting, this also provides the option to wait, which tells the hooks framework to try invoking the 
+  hook again later, and rollback, which tells the framework to rollback to the previous store-version in all regions.
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
+
+1. Within the push job.
+2. Within the parent controller.
+3. Within the child controllers.
+
+It is also possible to consider invoking some hooks in one of these location while other hooks would be executed 
+elsewhere. The table below summarizes the feasibility and tradeoffs for each of the proposed hooks:
+
+|                                              | Child Controller | Parent Controller | Push Job |
+|---------------------------------------------:|:----------------:|:-----------------:|:--------:|
+|                          `preStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                         `postStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                      `preSchemaRegistration` |        ❌         |         ✅         |   1️⃣    |
+|                     `postSchemaRegistration` |        ✅         |         ✅         |   1️⃣    |
+|                    `preStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|                   `postStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|  `preStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+| `postStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+|          `postStoreVersionLeaderReplication` |        ✅         |         ❌         |    ❌     |
+|                        `preStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                       `postStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                            `preEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+|                           `postEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+
```

Comment:
This is covering a bigger scope than region-by-region version swap and are we supposed to support all of them?

Reply:
We don't necessarily need to support everything. As I mentioned later in the doc, I recommend skipping the replication complete hook, since that one is most easily accomplished as a child controller hook, and there are other reasons why I think that is not the most straightforward approach. If we could have everything, I think replication complete would be an interesting event to emit, but it is not strictly required in the short-term.

Conversely, there could be other hooks we want but which are currently omitted...

In other words, the exact scope of the first iteration can be tweaked, although many of these hooks are must haves in the short-term.

Reply:
I think the absolute must haves are:

```
preStartOfStoreVersionIngestionForDaVinci
preStoreVersionSwap
postStoreVersionSwap
preEndOfPushJob
```

And it would be cleaner (but not necessarily critical) to have those as well:

```
postStartOfPushJob
postSchemaRegistration
preStoreVersionCreation
postStoreVersionCreation
postEndOfPushJob
```

The rest are less important in the short-term.